hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/tensorflow/config/TensorBoardProcessMgr.java

```diff
@@ -0,0 +1,331 @@
+@DependsOn("Settings")
+public class TensorBoardProcessMgr {
+
+  private static final Logger logger = Logger.getLogger(TensorBoardProcessMgr.class.getName());
+
+  @EJB
+  private Settings settings;
+  @EJB
+  private HdfsUsersFacade hdfsUsersFacade;
+  @EJB
+  private HdfsLeDescriptorsFacade hdfsLeFacade;
+  @EJB
+  private HdfsUsersController hdfsUsersController;
+  @EJB
+  private TensorBoardFacade tensorBoardFacade;
+  @EJB
+  private ProjectFacade projectFacade;
+  @EJB
+  private UserFacade userFacade;
+
+  @PostConstruct
+  public void init() {
+  }
+
+  @PreDestroy
+  public void preDestroy() {
+
+  }
+  
+  @TransactionAttribute(TransactionAttributeType.NOT_SUPPORTED)
+  public TensorBoardDTO startTensorBoard(Project project, Users user, HdfsUsers hdfsUser, String hdfsLogdir)
+          throws IOException {
+
+    String prog = settings.getHopsworksDomainDir() + "/bin/tensorboard.sh";
+
+    Process process = null;
+    Integer port = 0;
+
+    BigInteger pid = null;
+
+    String tensorBoardDir = settings.getStagingDir() + Settings.TENSORBOARD_DIRS;
+
+    String localDir = DigestUtils.sha256Hex(project.getName() + "_" + hdfsUser.getName());
+    tensorBoardDir = tensorBoardDir + File.separator + localDir;
+
+    File tbDir = new File(tensorBoardDir);
+    if(tbDir.exists()) {
+      for(File file: tbDir.listFiles()) {
+        if(file.getName().endsWith(".pid")) {
+          String pidContents = com.google.common.io.Files.readFirstLine(file, Charset.defaultCharset());
+          try {
+            pid = BigInteger.valueOf(Long.parseLong(pidContents));
+            if (pid != null && ping(pid) == 0) {
+              killTensorBoard(pid);
+              try {
+                TensorBoard tb = tensorBoardFacade.findForProjectAndUser(project, user);
+                if(tb.getPid().equals(pid)) {
+                  tensorBoardFacade.remove(tb);
+                }
+              } catch(DatabaseException dbe) {
+                //skip
+              }
+            }
+          } catch(NumberFormatException nfe) {
+            logger.log(Level.WARNING, "Expected number in pidfile " +
+                    file.getAbsolutePath() + " got " + pidContents);
+          }
+        }
+      }
+      FileUtils.deleteDirectory(tbDir);
+    }
+    tbDir.mkdirs();
+
+    String anacondaEnvironmentPath = settings.getAnacondaProjectDir(project.getName());
+
+    int retries = 3;
+
+    while(retries > 0) {
+
+      if(retries == 0) {
+        throw new IOException("Failed to start TensorBoard for project=" + project.getName() + ", user="
+                + user.getUid());
+      }
+
+      // use pidfile to kill any running servers
+      port = ThreadLocalRandom.current().nextInt(40000, 59999);
+
+      String[] command = new String[]{"/usr/bin/sudo", prog, "start", hdfsUser.getName(), hdfsLogdir,
+        tensorBoardDir, port.toString(), anacondaEnvironmentPath, settings.getHadoopVersion()};
+
```

Comment:
You forgot to materialize the user certificates before launching the TensorBoard. If TLS is enabled and a user starts the TensorBoard immediately after the login, changes are that the certificates are not materialized on the machine.
TensorBoard will fail to read.
Depending on which user is running the TensorBoard you might have privilege issues when reading the certificates. If that is the case you can specify another directory to materialize the certificates and change the `tensorboard.sh` script to export the `MATERIAL_DIRECTORY` env variable.
Also remember that you have to "dematerialize" the certificates once you are done using them, i.e. when you stop the TensorBoard.

Reply:
Thanks I missed that, will add it too.